src/algebraic_geometry/structure_sheaf.lean

```diff
@@ -507,4 +496,278 @@ def stalk_iso (x : Spec.Top R) :
+lemma to_basic_open_surjective (f : R) : function.surjective (to_basic_open R f) :=
+begin
+  intro s,
+  -- In this proof, `basic_open f` will play two distinct roles: Firstly, it is an open set in the
+  -- prime spectrum. Secondly, it is used as an indexing type for various families of objects
+  -- (open sets, ring elements, ...). In order to make the distinction clear, we introduce a type
+  -- alias `ι` that is used whenever we want think of it as an indexing type.
+  let ι : Type u := basic_open f,
+
+  -- First, we pick some cover of basic opens, on which we can represent `s` as a fraction
+  choose a' h' iDh' hxDh' s_eq' using locally_const_basic_open R (basic_open f) s,
+  -- Since basic opens are compact, we can pass to a finite subcover
+  obtain ⟨t, ht_cover'⟩ := (is_compact_basic_open f).elim_finite_subcover
+   (λ (i : ι), (basic_open (h' i)).1) (λ i, is_open_basic_open) (λ x hx, _),
+  swap,
+  { -- Here, we need to show that our basic opens actually form a cover of `basic_open f`
+    rw set.mem_Union,
+    exact ⟨⟨x,hx⟩, hxDh' ⟨x, hx⟩⟩ },
+  -- We use the normalization lemma from above to obtain the relation `a i * h j = h i * a j`
+  obtain ⟨a, h, iDh, ht_cover, ah_ha, s_eq⟩ := normalize_finite_fraction_representation R
+    (basic_open f) s t a' h' iDh' ht_cover' s_eq',
+  clear s_eq' iDh' hxDh' ht_cover' a' h',
+  -- Next we show that some power of `f` is a linear combination of the `h i`
+  obtain ⟨n, hn⟩ : f ∈ (ideal.span (h '' ↑t)).radical,
+  { rw [← vanishing_ideal_zero_locus_eq_radical, zero_locus_span],
+    simp_rw [subtype.val_eq_coe, basic_open_eq_zero_locus_compl] at ht_cover,
+    rw set.compl_subset_comm at ht_cover, -- Why doesn't `simp_rw` do this?
+    simp_rw [set.compl_Union, compl_compl, ← zero_locus_Union, ← finset.set_bUnion_coe,
+             ← set.image_eq_Union ] at ht_cover,
+    apply vanishing_ideal_anti_mono ht_cover,
+    exact subset_vanishing_ideal_zero_locus {f} (set.mem_singleton f) },
+
+  replace hn := ideal.mul_mem_left _ f hn,
+  erw [←pow_succ, finsupp.mem_span_iff_total] at hn,
+  rcases hn with ⟨b, b_supp, hb⟩,
+  rw finsupp.total_apply_of_mem_supported R b_supp at hb,
+  dsimp at hb,
+
+  -- Finally, we have all the ingredients.
+  -- We claim that our preimage is given by `(∑ (i : ι) in t, b i * a i) / f ^ (n+1)`
+  use (localization.of (submonoid.powers f)).mk' (∑ (i : ι) in t, b i * a i) ⟨f ^ (n+1), n+1, rfl⟩,
+  rw to_basic_open_mk',
+
+  -- Since the structure sheaf is a sheaf, we can show the desired equality locally.
+  -- Annoyingly, `sheaf.eq_of_locally_eq` requires an open cover indexed by a *type*, so we need to
+  -- coerce our finset `t` to a type first.
+  let tt := ((t : set (basic_open f)) : Type u),
+  -- This would work a little nicer if we could write `(structure_sheaf R).eq_of_locally_eq'`.
+  -- For that, the API on unique gluing should be extended to work with more general (algebraic?)
+  -- categories, rather than only sheaves of `Type`'s
```

Comment:
It seems that as long as `mono f` implies that the underlying function is `injective`, you can prove `eq_of_locally_eq` for an arbitrary `concrete_category`. This isn't universally true, but pretty close: whenever the forgetful functor to `Type` is a right adjoint.

I don't think we have an existing typeclass for that, but perhaps it's worth having one. Perhaps @b-mehta's `representably_concrete` on #7346 is exactly what we need?

Reply:
I think you mean #7436?
Yes, that's exactly what we would need here. However, if we were to develop sheaves of algebraic structures more generally, I think we want more structure on the category than just having a left adjoint to the forgetful functor. Maybe something along the lines of #5927, where the forgetful functor also preserves limits and reflects isomorphisms. Then we could also have *existence* of gluings, from locally defined sections (instead of only uniqueness, as in `eq_of_locally_eq`).
Maybe we should have a typeclass of this stronger notion? Or something like "algebraic category", as defined in [nlab](https://ncatlab.org/nlab/show/algebraic+category). I think this would be the most natural framework for "hands-on" sheaves.

Reply:
I think a point of consideration here is that it's sometimes helpful to work with sheaves in categories which aren't algebraic, for instance sheaves of topological rings: these are representably concrete but not algebraic. In the design in #5927, the general notion of sheaf makes sense in these cases as well, and when the forgetful functor has the additionally nice requirements like preserving limits and reflecting isomorphisms, we get a further more convenient notion. 
So I think Scott's idea of using representably concrete (which is a weaker condition than forgetful having a left adjoint) is a good one, and if for particular algebraic categories it turns out helpful to make stronger assumptions, we can do that at a later point?

Reply:
Sounds good!